### ISSUE TITLE
Discussion: How to handle events

### ISSUE THREAD

**Issue by kevzhao2** [NyxStudios/Orion#54]
Currently, there are `EventHandlerCollection<TEventArgs>` properties on all of the services. To register an event, you'd do something like the following:

```c#
private readonly Lazy<PlayerService> _playerService;

// ...

_playerService.Value.PlayerJoin.RegisterHandler(PlayerJoinHandler, Log);

// ...

[EventHandler(EventPriority.Monitor)]
public void PlayerJoinHandler(object? sender, PlayerJoinEventArgs e) {
    Log.Debug("{Player} joined", e.Player);
}
```

However, another possible way to register events could be like the following:

```c#
Kernel.RegisterEvents(this);

// ...

[EventHandler(EventPriority.Monitor)]
public void OnPlayerJoin(PlayerJoinEventArgs e) {
    Log.Debug("{Player} joined", e.Player);
}

[EventHandler(EventPriority.Monitor)]
public void OnPlayerQuit(PlayerQuitEventArgs e) {
    Log.Debug("{Player} quit", e.Player);
}
```

The idea is that `OrionKernel.RegisterEvents` will take all methods from the given object, filtering using the ones with `EventHandlerAttribute` annotating them, and register them automatically. There would be no requirement to define any `EventHandlerCollection`s anywhere, they'd just get aggregated in the kernel. Additionally, to invoke the events, you would just do the following:

```c#
Kernel.InvokeEvent(new PlayerJoinEventArgs(player));
Kernel.InvokeEvent(new PlayerQuitEventArgs(player));
```

TL:DR; first approach is explicit, second approach is more similar to Bukkit's. What are your thoughts on these two approaches?

**Comment by QuiCM** [comment]
First looks much more verbose and intuitive and thus, for a user unfamiliar with the internals, more understandable (imo). If a handler is not picking up events correctly it is very easy to understand why without having to look under the hood.

The second looks cleaner and simpler and is my preference, however looks less intuitive and more difficult to debug without knowledge of the parent system

**Comment by kevzhao2** [comment]
Just to update this, the direction forward has been chosen: aggregating events and event handlers in the kernel.
